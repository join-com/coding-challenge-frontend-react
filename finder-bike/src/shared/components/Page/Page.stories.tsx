import { storiesOf } from '@storybook/react';
import React, { useState, useCallback } from 'react';
import Page from './Page';
import { action } from '@storybook/addon-actions';

function PagePlayGround() {
  const [currentPage, setCurrentPage] = useState(1);
  const onClick = useCallback(value => {
    setCurrentPage(value);
  }, []);

  return (
    <Page
      pageSize={10}
      currentPage={currentPage}
      totalItems={100}
      maxPages={5}
      onClick={onClick}
    />
  );
}

storiesOf('Components/Page', module)
  .add('Basic', () => (
    <Page
      pageSize={10}
      currentPage={4}
      totalItems={100}
      maxPages={5}
      onClick={action('onClick')}
    />
  ))
  .add('Playground', () => <PagePlayGround />);
