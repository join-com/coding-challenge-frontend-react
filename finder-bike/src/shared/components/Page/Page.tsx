import React, { useCallback } from 'react';
import styled from 'styled-components';
import { Text } from '../../ui-kits/Text';
import * as Colors from '../../ui-kits/Variables/Colors';
import paginate from './paginate';
import { Icon } from '../../ui-kits/Icon';
import {
  faChevronLeft,
  faChevronRight
} from '@fortawesome/free-solid-svg-icons';

export interface IPage {
  currentPage?: number;
  pageSize?: number;
  totalItems: number;
  maxPages?: number;
  onClick: (e: any) => void;
}

const StyledPageWrapper = styled.div`
  width: 100%;
  margin: 1rem 0;
`;

const StyledPageList = styled.div`
  display: grid;
  grid-template-columns: repeat(auto-fit, minmax(50px, 1fr));
  grid-template-rows: 1fr;
  text-align: center;
`;

const StyledPageListItem = styled.a`
  position: relative;
  display: block;
  padding: 0.75rem 1.25rem;
  margin-left: -1px;
  line-height: 1.25;

  border: 1px solid ${Colors.grey2};

  ${({ selected }) =>
    selected
      ? `
    background-color: ${Colors.primary};
    color: ${Colors.white}
  `
      : `
    background-color: ${Colors.white};
    color: ${Colors.primary}
  `}

  ${({ disabled, selected }) =>
    disabled
      ? `
    background-color: ${Colors.white};
    color: ${Colors.grey}
  `
      : !selected &&
        `
    cursor: pointer;
    :hover {
      text-decoration: none;
      background-color: ${Colors.grey};
    }
  `}
`;

function Page(props: IPage) {
  const { currentPage, totalPages, pages } = paginate(
    props.totalItems,
    props.currentPage,
    props.pageSize,
    props.maxPages
  );
  const selectPage = useCallback(
    value => {
      props.onClick(value);
    },
    [props]
  );

  return (
    <StyledPageWrapper>
      <StyledPageList>
        <StyledPageListItem
          disabled={pages[0] === 1}
          onClick={() => selectPage(currentPage - 1)}
        >
          <Icon size={'2x'} icon={faChevronLeft} />
        </StyledPageListItem>

        {pages.map(page => (
          <StyledPageListItem
            selected={currentPage === page}
            key={page}
            onClick={() => selectPage(page)}
          >
            <Text>{page}</Text>
          </StyledPageListItem>
        ))}

        <StyledPageListItem
          disabled={pages[pages.length - 1] === totalPages}
          onClick={() => selectPage(currentPage + 1)}
        >
          <Icon size={'2x'} icon={faChevronRight} />
        </StyledPageListItem>
      </StyledPageList>
    </StyledPageWrapper>
  );
}

export default Page;
