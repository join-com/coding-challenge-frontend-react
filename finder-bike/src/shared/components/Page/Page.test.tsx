import React from 'react';
import { render } from '@testing-library/react';
import Page from './Page';

describe('<Page /> spec', () => {
  it('renders the component', () => {
    const component = render(
      <Page
        pageSize={10}
        currentPage={5}
        totalItems={100}
        maxPages={5}
        onClick={jest.fn}
      />
    );
    expect(component).toMatchSnapshot();
  });
});
