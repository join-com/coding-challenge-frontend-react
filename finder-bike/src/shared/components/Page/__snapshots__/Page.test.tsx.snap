// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Page /> spec renders the component 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="sc-bxivhb MsVIk"
      >
        <div
          class="sc-ifAKCX jpirIC"
        >
          <a
            class="sc-EHOje iAEShO"
          >
            <svg
              aria-hidden="true"
              class="svg-inline--fa fa-chevron-left fa-w-10 fa-2x sc-htpNat KtFsv"
              data-icon="chevron-left"
              data-prefix="fas"
              focusable="false"
              role="img"
              viewBox="0 0 320 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M34.52 239.03L228.87 44.69c9.37-9.37 24.57-9.37 33.94 0l22.67 22.67c9.36 9.36 9.37 24.52.04 33.9L131.49 256l154.02 154.75c9.34 9.38 9.32 24.54-.04 33.9l-22.67 22.67c-9.37 9.37-24.57 9.37-33.94 0L34.52 272.97c-9.37-9.37-9.37-24.57 0-33.94z"
                fill="currentColor"
              />
            </svg>
          </a>
          <a
            class="sc-EHOje iAEShO"
          >
            <span
              class="sc-bdVaJa iIMtwm"
            >
              3
            </span>
          </a>
          <a
            class="sc-EHOje iAEShO"
          >
            <span
              class="sc-bdVaJa iIMtwm"
            >
              4
            </span>
          </a>
          <a
            class="sc-EHOje eLtbHC"
          >
            <span
              class="sc-bdVaJa iIMtwm"
            >
              5
            </span>
          </a>
          <a
            class="sc-EHOje iAEShO"
          >
            <span
              class="sc-bdVaJa iIMtwm"
            >
              6
            </span>
          </a>
          <a
            class="sc-EHOje iAEShO"
          >
            <span
              class="sc-bdVaJa iIMtwm"
            >
              7
            </span>
          </a>
          <a
            class="sc-EHOje iAEShO"
          >
            <svg
              aria-hidden="true"
              class="svg-inline--fa fa-chevron-right fa-w-10 fa-2x sc-htpNat KtFsv"
              data-icon="chevron-right"
              data-prefix="fas"
              focusable="false"
              role="img"
              viewBox="0 0 320 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M285.476 272.971L91.132 467.314c-9.373 9.373-24.569 9.373-33.941 0l-22.667-22.667c-9.357-9.357-9.375-24.522-.04-33.901L188.505 256 34.484 101.255c-9.335-9.379-9.317-24.544.04-33.901l22.667-22.667c9.373-9.373 24.569-9.373 33.941 0L285.475 239.03c9.373 9.372 9.373 24.568.001 33.941z"
                fill="currentColor"
              />
            </svg>
          </a>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="sc-bxivhb MsVIk"
    >
      <div
        class="sc-ifAKCX jpirIC"
      >
        <a
          class="sc-EHOje iAEShO"
        >
          <svg
            aria-hidden="true"
            class="svg-inline--fa fa-chevron-left fa-w-10 fa-2x sc-htpNat KtFsv"
            data-icon="chevron-left"
            data-prefix="fas"
            focusable="false"
            role="img"
            viewBox="0 0 320 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M34.52 239.03L228.87 44.69c9.37-9.37 24.57-9.37 33.94 0l22.67 22.67c9.36 9.36 9.37 24.52.04 33.9L131.49 256l154.02 154.75c9.34 9.38 9.32 24.54-.04 33.9l-22.67 22.67c-9.37 9.37-24.57 9.37-33.94 0L34.52 272.97c-9.37-9.37-9.37-24.57 0-33.94z"
              fill="currentColor"
            />
          </svg>
        </a>
        <a
          class="sc-EHOje iAEShO"
        >
          <span
            class="sc-bdVaJa iIMtwm"
          >
            3
          </span>
        </a>
        <a
          class="sc-EHOje iAEShO"
        >
          <span
            class="sc-bdVaJa iIMtwm"
          >
            4
          </span>
        </a>
        <a
          class="sc-EHOje eLtbHC"
        >
          <span
            class="sc-bdVaJa iIMtwm"
          >
            5
          </span>
        </a>
        <a
          class="sc-EHOje iAEShO"
        >
          <span
            class="sc-bdVaJa iIMtwm"
          >
            6
          </span>
        </a>
        <a
          class="sc-EHOje iAEShO"
        >
          <span
            class="sc-bdVaJa iIMtwm"
          >
            7
          </span>
        </a>
        <a
          class="sc-EHOje iAEShO"
        >
          <svg
            aria-hidden="true"
            class="svg-inline--fa fa-chevron-right fa-w-10 fa-2x sc-htpNat KtFsv"
            data-icon="chevron-right"
            data-prefix="fas"
            focusable="false"
            role="img"
            viewBox="0 0 320 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M285.476 272.971L91.132 467.314c-9.373 9.373-24.569 9.373-33.941 0l-22.667-22.667c-9.357-9.357-9.375-24.522-.04-33.901L188.505 256 34.484 101.255c-9.335-9.379-9.317-24.544.04-33.901l22.667-22.667c9.373-9.373 24.569-9.373 33.941 0L285.475 239.03c9.373 9.372 9.373 24.568.001 33.941z"
              fill="currentColor"
            />
          </svg>
        </a>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
